src/sardana/macroserver/scan/scandata.py

```diff
@@ -186,6 +189,67 @@ def setComplete(self):
     def setWritten(self):
         self.written = 1
 
+    def __get_fullname(self, proxy, scheme_implicit=True, t3_style=False):
+        """
+        :param proxy: taurus
+        :param t3_style:
+        :return:
+        """
+        v = proxy.getNameValidator()
+        params = v.getParams(proxy.getFullName())
+        if t3_style:
+
+            fullname = '{0}:{1}/{2}'.format(params['host'].split('.')[0],
+                                            params['port'],
+                                            params['devicename'])
+        else:
+            fullname = '{0}:{1}/{2}'.format(socket.getfqdn(params['host']),
+                                            params['port'],
+                                            params['devicename'])
+            if scheme_implicit:
+                fullname = 'tango://{0}'.format(fullname)
+
+        attr_name = params.get('attrubutename', None)
+        if attr_name is not None:
+            fullname = '{0}/{1}'.format(fullname, params['attributename'])
+
+        return fullname
+
+    def __getitem__(self, item):
+        item = item.lower()
+        if item == 'dt':
+            item = 'timestamp'
+        if item in self.data:
+            return self.data[item]
+
+        # check if the item is a device name or an attribute name
+        try:
+            chn_proxy = Device(item)
+        except TaurusException:
+            try:
+                chn_proxy = Attribute(item)
+            except TaurusException:
+                raise KeyError(item)
+
+        # TODO: Refactor this code when the data dictionary uses only taurus
+        # 4 full name
+
+        try:
```

Comment:
My suggestion is to refactor the whole thing as follows:
- remove  `__get_fullnames`
- create `__get_t3_name`  that does the t3-specific part of `__get_fullnames`
- create `__get_t4_name` that finds (without instantiating anything) whether the item is an attr or dev, and  then uses `validator.getNames(item)[0]` for fnding the full name
- create `__get_tango_name` (to support the special case of tango-urls without scheme name) and implement it as 

  ```
  n = __get_t4_name()
  if n startswith('tango://'):
      n = n[8:]
  return n
  ```

- Finally refactor this try block of ` __getitem__` to sequentially construct names with the above functions and try to get them from `data`